File: lib/python/opentoken-cli/src/main/opentoken_cli/commands/open_token_command.py
"""
Copyright (c) Truveta. All rights reserved.
"""

import argparse
import logging
import os
import sys

from opentoken.metadata import Metadata
from opentoken_cli.util.version_checker import start_version_check

logger = logging.getLogger(__name__)


class OpenTokenCommand:
    """
    Main entry point command for OpenToken CLI with subcommands.
    Provides modern, subcommand-based interface for token operations.
    """

    VERSION = Metadata.DEFAULT_VERSION

    @staticmethod
    def show_banner():
        """
        Display the OpenToken banner for interactive sessions.
        Respects NO_COLOR environment variable and TTY detection.
        """
        # Check if we're in an interactive terminal and NO_COLOR is not set
        if not OpenTokenCommand._is_interactive() or os.getenv("NO_COLOR"):
            return

        try:
            banner = OpenTokenCommand._get_colorized_banner()
            print(banner)
        except Exception as e:
            # Silently fail banner display - it's not critical
            logger.debug(f"Could not display banner: {e}")

    @staticmethod
    def _is_interactive():
        """Check if stdout is connected to an interactive terminal."""
        return sys.stdout.isatty()

    @staticmethod
    def _get_colorized_banner():
        """Get the colorized OpenToken banner."""
        cyan = "\033[36m"
        blue = "\033[34m"
        reset = "\033[0m"

        return (
            f"{cyan}  ___                 _____     _              {reset}\n"
            f"{cyan} / _ \\ _ __   ___ _ _|_   _|__ | | _____ _ __  {reset}\n"
            f"{cyan}| | | | '_ \\ / _ \\ '_ \\| |/ _ \\| |/ / _ \\ '_ \\ {reset}\n"
            f"{cyan}| |_| | |_) |  __/ | | | | (_) |   <  __/ | | |{reset}\n"
            f"{cyan} \\___/| .__/ \\___|_| |_|_|\\___/|_|\\_\\___|_| |_|{reset}\n"
            f"{cyan}      |_|                                       {reset}\n"
            f"{blue}Privacy-Preserving Record Linkage v{OpenTokenCommand.VERSION}{reset}\n"
        )

    @staticmethod
    def create_parser():
        """Create the main argument parser with subcommands."""
        parser = argparse.ArgumentParser(
            prog="opentoken",
            description="Privacy-preserving record linkage via cryptographic tokens",
            formatter_class=argparse.RawDescriptionHelpFormatter,
        )

        parser.add_argument(
            "--version",
            action="version",
            version=f"OpenToken {OpenTokenCommand.VERSION}",
        )

        parser.add_argument(
            "--no-update-check",
            action="store_true",
            default=False,
            dest="no_update_check",
            help="Disable the automatic update check on startup",
        )

        subparsers = parser.add_subparsers(
            title="commands",
            description="Available commands",
            dest="command",
            help="Use 'opentoken <command> --help' for command-specific help",
        )

        # Import command modules here to avoid circular imports
        from opentoken_cli.commands.decrypt_command import DecryptCommand
        from opentoken_cli.commands.encrypt_command import EncryptCommand
        from opentoken_cli.commands.extension_command import ExtensionCommand
        from opentoken_cli.commands.generate_key_pair_command import GenerateKeyPairCommand
        from opentoken_cli.commands.help_command import HelpCommand
        from opentoken_cli.commands.initiate_exchange_command import InitiateExchangeCommand
        from opentoken_cli.commands.package_command import PackageCommand
        from opentoken_cli.commands.tokenize_command import TokenizeCommand
        from opentoken_cli.commands.update_command import UpdateCommand
        from opentoken_cli.extension.extension_loader import BUILTIN_COMMANDS, ExtensionLoader

        # Register subcommands in alphabetical order by command name
        DecryptCommand.register_subcommand(subparsers)
        EncryptCommand.register_subcommand(subparsers)
        ExtensionCommand.register_subcommand(subparsers)
        GenerateKeyPairCommand.register_subcommand(subparsers)
        HelpCommand.register_subcommand(subparsers)
        InitiateExchangeCommand.register_subcommand(subparsers)
        PackageCommand.register_subcommand(subparsers)
        TokenizeCommand.register_subcommand(subparsers)
        UpdateCommand.register_subcommand(subparsers)

        # Load installed extensions (entry points or registry).
        # BUILTIN_COMMANDS is the authoritative set of reserved command names.
        ExtensionLoader.load_extensions(subparsers, BUILTIN_COMMANDS)

        # Sort all subcommands (built-ins and extensions) alphabetically.
        # Uses private argparse internals that may not exist in all Python versions;
        # sorting is skipped gracefully if those attributes are absent.
        if hasattr(subparsers, "_name_parser_map") and hasattr(subparsers, "_choices_actions"):
            sorted_items = sorted(subparsers._name_parser_map.items())
            subparsers._name_parser_map.clear()
            subparsers._name_parser_map.update(sorted_items)
            subparsers._choices_actions.sort(key=lambda a: a.dest)
        else:
            logger.debug(
                "Subcommand alphabetical sorting skipped: argparse internals "
                "(_name_parser_map / _choices_actions) not available on this Python version."
            )

        return parser

    @staticmethod
    def main(args=None):
        """Main entry point for the command-line application."""
        parser = OpenTokenCommand.create_parser()

        # Show banner for interactive runs.
        OpenTokenCommand.show_banner()

        try:
            parsed_args = parser.parse_args(args)
        except SystemExit as error:
            return error.code if isinstance(error.code, int) else 1

        no_update_check = getattr(parsed_args, "no_update_check", False)
        should_start_version_check = OpenTokenCommand._should_start_version_check(parsed_args)

        version_checker = None
        if should_start_version_check:
            # Start the asynchronous version check before executing the command
            version_checker = start_version_check(OpenTokenCommand.VERSION, no_update_check=no_update_check)

        # If no subcommand specified, show help
        if not parsed_args.command:
            parser.print_help()
            if version_checker is not None:
                version_checker.wait_and_notify()
            return 0

        # Execute the command
        try:
            exit_code = parsed_args.func(parsed_args)
        except Exception as e:
            logger.error(f"Command execution failed: {e}", exc_info=True)
            exit_code = 1

        # Wait for the version check and surface any update notice after command output
        if version_checker is not None:
            version_checker.wait_and_notify()
        return exit_code

    @staticmethod
    def execute(args):
        """
        Execute the CLI without calling sys.exit().
        Useful for testing or when embedding the CLI in another application.

        Args:
            args: Command-line arguments as a list

        Returns:
            Exit code (0 for success, non-zero for errors)
        """
        return OpenTokenCommand.main(args)

    @staticmethod
    def _is_help_request(args):
        """Check if the command is a help request."""
        if not args:
            return False
        for arg in args:
            if arg in ("--help", "help"):
                return True
        return False

    @staticmethod
    def _should_start_version_check(parsed_args: argparse.Namespace) -> bool:
        """Return whether startup version checks should run for the parsed command."""
        return getattr(parsed_args, "command", None) != "update"

File: lib/java/opentoken-cli/src/main/java/com/truveta/opentoken/cli/commands/OpenTokenCommand.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.cli.commands;

import java.util.concurrent.Callable;

import org.jline.terminal.Terminal;
import org.jline.terminal.TerminalBuilder;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.truveta.opentoken.Metadata;
import com.truveta.opentoken.cli.util.VersionChecker;

import picocli.CommandLine;
import picocli.CommandLine.Command;
import picocli.CommandLine.HelpCommand;
import picocli.CommandLine.Option;

/**
 * Main entry point command for OpenToken CLI with subcommands.
 * Provides modern, subcommand-based interface for token operations.
 */
@Command(name = "opentoken", description = "Privacy-preserving record linkage via cryptographic tokens", mixinStandardHelpOptions = true, version = OpenTokenCommand.CLI_VERSION, subcommands = {
        HelpCommand.class,
        TokenizeCommand.class,
        EncryptCommand.class,
        DecryptCommand.class,
        PackageCommand.class,
        UpdateCommand.class
})
public class OpenTokenCommand implements Callable<Integer> {

    private static final String VERSION = Metadata.DEFAULT_VERSION;
    public static final String CLI_VERSION = "OpenToken " + VERSION;
    private static final String BANNER_SUBTITLE = "Privacy-Preserving Record Linkage v" + VERSION;

    private static final Logger logger = LoggerFactory.getLogger(OpenTokenCommand.class);

    @Option(names = { "--no-update-check" }, description = "Disable the automatic update check on startup")
    private boolean noUpdateCheck;

    /**
     * Display the OpenToken banner for interactive sessions.
     */
    public static void showBanner() {
        // Check if we're in an interactive terminal and NO_COLOR is not set
        if (!isInteractive() || System.getenv("NO_COLOR") != null) {
            return;
        }

        try {
            Terminal terminal = TerminalBuilder.builder()
                    .system(true)
                    .build();

            String banner = getColorizedBanner();
            terminal.writer().println(banner);
            terminal.writer().flush();
        } catch (Exception e) {
            // Silently fail banner display - it's not critical
            logger.debug("Could not display banner", e);
        }
    }

    /**
     * Check if stdout is connected to an interactive terminal.
     */
    private static boolean isInteractive() {
        try {
            Terminal terminal = TerminalBuilder.builder()
                    .system(true)
                    .dumb(true)
                    .build();
            return terminal.getType() != Terminal.TYPE_DUMB;
        } catch (Exception e) {
            return false;
        }
    }

    /**
     * Get the colorized OpenToken banner.
     */
    private static String getColorizedBanner() {
        String cyan = "\u001B[36m";
        String blue = "\u001B[34m";
        String reset = "\u001B[0m";

        return cyan + "  ___                 _____     _              " + reset + "\n" +
                cyan + " / _ \\ _ __   ___ _ _|_   _|__ | | _____ _ __  " + reset + "\n" +
                cyan + "| | | | '_ \\ / _ \\ '_ \\| |/ _ \\| |/ / _ \\ '_ \\ " + reset + "\n" +
                cyan + "| |_| | |_) |  __/ | | | | (_) |   <  __/ | | |" + reset + "\n" +
                cyan + " \\___/| .__/ \\___|_| |_|_|\\___/|_|\\_\\___|_| |_|" + reset + "\n" +
                cyan + "      |_|                                       " + reset + "\n" +
                blue + BANNER_SUBTITLE + reset + "\n";
    }

    @Override
    public Integer call() {
        // If no subcommand is specified, show help
        CommandLine.usage(this, System.out);
        return 0;
    }

    /**
     * Main entry point for the command-line application.
     */
    public static void main(String[] args) {
        int exitCode = execute(args);
        System.exit(exitCode);
    }

    /**
     * Execute the CLI without calling System.exit().
     * Useful for testing or when embedding the CLI in another application.
     *
     * @param args command-line arguments
     * @return exit code (0 for success, non-zero for errors)
     */
    public static int execute(String[] args) {
        // Show banner for interactive runs (not for --help or piped output)
        if (args.length == 0 || !isHelpRequest(args)) {
            showBanner();
        }

        OpenTokenCommand rootCommand = new OpenTokenCommand();
        CommandLine commandLine = new CommandLine(rootCommand);
        commandLine.setExecutionStrategy(new CommandLine.RunLast());

        CommandLine.ParseResult parseResult = null;
        try {
            parseResult = commandLine.parseArgs(args);
        } catch (CommandLine.ParameterException e) {
            logger.debug("Could not parse command args before startup update check", e);
        }

        // Determine --no-update-check before full parse (peek at args)
        boolean noUpdateCheck = hasFlag(args, "--no-update-check");
        boolean shouldRunStartupVersionCheck = shouldRunStartupVersionCheck(parseResult);

        VersionChecker versionChecker = new VersionChecker(VERSION, noUpdateCheck);
        if (shouldRunStartupVersionCheck) {
            versionChecker.start();
        }

        int exitCode = parseResult == null
                ? commandLine.execute(args)
                : commandLine.getExecutionStrategy().execute(parseResult);

        if (shouldRunStartupVersionCheck) {
            versionChecker.waitAndNotify();
        }

        return exitCode;
    }

    /**
     * Return whether the startup update check should run for the parsed command.
     *
     * @param parseResult the parsed command result
     * @return {@code true} for non-update commands; {@code false} for update
     */
    static boolean shouldRunStartupVersionCheck(CommandLine.ParseResult parseResult) {
        if (parseResult == null || parseResult.subcommand() == null) {
            return true;
        }
        return !"update".equals(parseResult.subcommand().commandSpec().name());
    }

    /**
     * Check if the command is a help request.
     */
    private static boolean isHelpRequest(String[] args) {
        for (String arg : args) {
            if (arg.equals("--help") || arg.equals("help")) {
                return true;
            }
        }
        return false;
    }

    /**
     * Check whether a specific flag is present in the argument list.
     *
     * @param args the argument array
     * @param flag the flag to look for
     * @return {@code true} if the flag is present
     */
    private static boolean hasFlag(String[] args, String flag) {
        for (String arg : args) {
            if (arg.equals(flag)) {
                return true;
            }
        }
        return false;
    }
}
